packages/core/src/browser/frontend-connection-status.ts

```diff
@@ -0,0 +1,221 @@
+            };
+            xhr.onerror = () => {
+                this.connectionState = this.connectionState.next(false);
+                this.contributionProvider.getContributions().forEach(contribution => contribution.onStatusChange(this.connectionState));
+            };
+            xhr.open('GET', `${window.location.href}alive`);
+            try { xhr.send(); } catch { /* NOOP */ }
+        }, this.options.retryInterval);
+        this.contributionProvider.getContributions().forEach(contribution => contribution.onStatusChange(this.connectionState));
+    }
+
+    onStop() {
+        if (this.timer !== undefined) {
+            clearInterval(this.timer);
+            this.timer = undefined;
+        }
+    }
+
+}
+
+@injectable()
+export class ConnectionStatusStatusBarContribution implements ConnectionStatusContribution {
+
+    constructor( @inject(StatusBar) protected statusBar: StatusBar) {
+    }
+
+    onStatusChange(event: ConnectionStatusChangeEvent) {
+        this.statusBar.removeElement('connection-status');
+        const text = `$(${this.getStatusIcon(event.health)})`;
+        const tooltip = event.health ? `Connection health: ${event.health}%` : 'Not connected';
+        this.statusBar.setElement('connection-status', {
+            alignment: StatusBarAlignment.RIGHT,
+            text,
+            priority: 0,
+            tooltip
+        });
+    }
+
+    private getStatusIcon(health: number | undefined) {
+        if (health === undefined || health === 0) {
+            return 'exclamation-circle';
+        }
+        if (health < 25) {
+            return 'frown-o';
+        }
+        if (health < 50) {
+            return 'meh-o';
+        }
+        return 'smile-o';
+    }
+
+}
+
+export class ConnectionStateMachine implements ConnectionStatusChangeEvent {
+
+    static readonly MAX_HISTORY = 100;
+
+    public readonly health: number;
+
+    constructor(
+        private readonly props: { readonly threshold: number },
+        public readonly state: ConnectionState = ConnectionState.CONNECTED,
+        private readonly history: boolean[] = []) {
+
+        if (this.state === ConnectionState.CONNECTION_LOST) {
+            this.health = 0;
+        } else {
+            this.health = this.history.length === 0 ? 100 : Math.round((this.history.filter(success => success).length / this.history.length) * 100);
+        }
+    }
+
+    next(success: boolean): ConnectionStateMachine {
+        const newHistory = this.updateHistory(success);
+        // Initial optimism.
+        let hasConnection = true;
+        if (newHistory.length > this.props.threshold) {
+            hasConnection = newHistory.slice(-this.props.threshold).some(s => s);
+        }
+        // Ideally, we do not switch back to online if we see any `true` items but, let's say, after three consecutive `true`s.
+        return new ConnectionStateMachine(this.props, hasConnection ? ConnectionState.CONNECTED : ConnectionState.CONNECTION_LOST, newHistory);
```

Comment:
Do we need a new object each 2 sec?

Reply:
I do not understand your concern with this; we would create a new event instance anyway.

Reply:
ok